mmv1/third_party/terraform/services/composer/resource_composer_environment.go.erb

```diff
@@ -2812,7 +2823,60 @@ func versionsEqual(old, new string) (bool, error) {
 	return o.Equal(n), nil
 }
 
-func isComposer3(d *schema.ResourceData, config *transport_tpg.Config)	bool {
-	image_version := d.Get("config.0.software_config.0.image_version").(string)
-	return strings.Contains(image_version, "composer-3")
+func isComposer3(v interface{})	bool {
+	var imageVersion string
+	switch v.(type) {
+	case *schema.ResourceData:
+		imageVersion = v.(*schema.ResourceData).Get("config.0.software_config.0.image_version").(string)
+	case *schema.ResourceDiff:
+		imageVersion = v.(*schema.ResourceDiff).Get("config.0.software_config.0.image_version").(string)
+	case *composer.EnvironmentConfig:
+		imageVersion = v.(*composer.EnvironmentConfig).SoftwareConfig.ImageVersion
+	case string:
+		imageVersion = v.(string)
+	default:
+		log.Printf("[ERROR] error in plugin: unexpected argument type in isComposer3")
+	}
+	return strings.Contains(imageVersion, "composer-3")
+}
+
+func imageVersionChangeValidationFunc(ctx context.Context, old, new, meta any) error {
```

Comment:
Please add unit tests for this function.

Reply:
TestAccComposerEnvironmentComposer3_upgrade_expectError is testing this functionality. Do we still need additional tests?

Reply:
responded on the thread for the customizediff - TLDR I'm fine to stick with the acceptance tests for now at least.